src/validator.ts

```diff
@@ -2,49 +2,243 @@ import Ajv from 'ajv';
+        : 'unknown';
+    });
+  if (missingPropertyNames.length > 0) {
+    const missingMessage = prepend(
+      subject,
+      `must have required ${
+        missingPropertyNames.length > 1 ? 'properties' : 'property'
+      }: ${missingPropertyNames.join(', ')}.`
+    );
+    messageParts.push(missingMessage);
+  }
+};
+
+const typeErrors = (
+  subject: string,
+  errors: Array<ErrorObject>,
+  messageParts: Array<string>
 ) => {
+  const typeErrorsList: Array<string> = [];
+  let errorCount = 0;
+
+  for (let i = 0; i < errors.length; i++) {
+    const e = errors[i];
+
+    if (e.keyword === 'type') {
+      errorCount += 1;
+      if (errorCount > maxErrors) {
+        continue;
+      } else {
+        formatIndividualError(e, typeErrorsList);
+      }
+    }
+  }
+
+  if (errorCount > maxErrors) {
+    typeErrorsList.push(`and ${errorCount - maxErrors} other errors`);
+  }
+
+  if (typeErrorsList.length > 0) {
+    const prefix =
+      messageParts.length > 0
+        ? 'There were also type errors:'
+        : `${subject} had type errors:`;
+    const typeErrorMessage = prepend(prefix, typeErrorsList.join(', ')) + '.';
+
+    messageParts.push(typeErrorMessage);
+  }
+};
+
+const validationErrors = (
+  subject: string,
+  errors: Array<ErrorObject>,
+  messageParts: Array<string>
+) => {
+  const validationErrors: Array<string> = [];
+  let errorCount = 0;
+
+  // List of error keywords from https://ajv.js.org/api.html#validation-errors
+  for (let i = 0; i < errors.length; i++) {
+    const e = errors[i];
+
+    if (e.keyword === 'minLength' && e.params.limit === 1) {
+      e.message = 'must not be blank';
+    }
+
+    switch (e.keyword) {
+      case 'minimum':
+      case 'maximum':
+      case 'exclusiveMinimum':
+      case 'exclusiveMaximum':
+      case 'minLength':
+      case 'maxLength':
+      case 'maxProperties':
+      case 'minProperties':
+      case 'minItems':
+      case 'maxItems':
+      case 'additionalItems':
+      case 'additionalProperties':
+        errorCount += 1;
+        if (errorCount > maxErrors) {
+          continue;
+        } else {
+          formatIndividualError(e, validationErrors);
+        }
+        break;
+      default:
+      // noop, other non-validation error handled elsewhere
+    }
+  }
+
+  if (errorCount > maxErrors) {
+    validationErrors.push(`and ${errorCount - maxErrors} other errors`);
+  }
+
+  if (validationErrors.length > 0) {
+    const prefix =
+      messageParts.length > 0
+        ? 'There were also validation errors:'
+        : `${subject} had validation errors:`;
+    const validationErrorMessage =
+      prepend(prefix, validationErrors.join(', ')) + '.';
+
+    messageParts.push(validationErrorMessage);
+  }
+};
+
+export const errorFormatter = (subject: string, errors: Array<ErrorObject>) => {
+  const anyOfErrors = errors.filter(
+    (error) =>
+      error.schemaPath.indexOf('anyOf') > -1 && error.keyword !== 'anyOf'
```

Comment:
Ah, ok maybe I understand a bit more after seeing the unit test added in the query PR: https://github.com/pinecone-io/pinecone-ts-client/blob/80fae53150c1e9a31f9c9c9d205546f88e59ff82/src/__tests__/validator.test.ts#L506

Seems like there's an additional "anyOf" error added to the list?

Reply:
Yeah, there's an entry in the errors list just letting you know you're in an anyOf error situation, but it doesn't contain any content that needs to be shown to the user. There's a test case covering this situation where the AJV errors look like this: 

```
      const validationErrors = [
        {
          instancePath: '',
          schemaPath: '#/anyOf/0/required',
          keyword: 'required',
          params: { missingProperty: 'id' },
          message: "must have required property 'id'",
        },
        {
          instancePath: '',
          schemaPath: '#/anyOf/1/required',
          keyword: 'required',
          params: { missingProperty: 'vector' },
          message: "must have required property 'vector'",
        },
        {
          instancePath: '',
          schemaPath: '#/anyOf',
          keyword: 'anyOf',
          params: {},
          message: 'must match a schema in anyOf',
        },
      ];
```

So the thinking here is to disregard the item with `error.keyword === 'anyOf'` (since it doesn't have any content that ends up in the final message), then sort these errors by schema group, strip out the part of the schema path referring to anyOf (to avoid an infinite loop with the recursive call), and then make a recursive call to build the error message that should be shown for each error group. In this example there is only one error from each schema group, but it could be more complex if the applicable types are larger and more complicated.